File: native/holistic/android/HolisticFrameProcessorPlugin.kt
package com.app.signlanguagemobile.holistic

import android.graphics.Bitmap
import android.util.Log
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.holisticlandmarker.HolisticLandmarker
import com.google.mediapipe.tasks.vision.holisticlandmarker.HolisticLandmarkerResult
import android.media.Image
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.SharedArray
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Runs MediaPipe HolisticLandmarker over each camera frame and returns one
 * packed Float32 buffer.
 *
 * The layout is specified and unit-tested in `lib/extractors/holistic-buffer.ts`,
 * which is the authority for this encoding:
 *   [0]     schema version (1)
 *   [1]     timestamp ms since capture start
 *   [2..5]  face / pose / leftHand / rightHand counts, 0 when undetected
 *   then    each stream in that order, four floats per landmark: x, y, z, visibility
 *
 * A packed buffer is used rather than a map because a holistic result is ~543
 * landmarks; bridging that as nested objects every frame would dominate the
 * frame budget.
 *
 * Nothing here retains, encodes, or writes a frame.
 */
class HolisticFrameProcessorPlugin(
  private val proxy: VisionCameraProxy,
  options: Map<String, Any>?,
) : FrameProcessorPlugin() {

  private companion object {
    const val TAG = "HolisticPlugin"
    const val SCHEMA_VERSION = 1f
    const val HEADER_FLOATS = 6
    const val FLOATS_PER_LANDMARK = 4
    const val MODEL_ASSET = "holistic_landmarker.task"
    const val BYTES_PER_PIXEL = 4
  }

  private val landmarker: HolisticLandmarker? = try {
    val base = BaseOptions.builder()
      .setModelAssetPath(MODEL_ASSET)
      .setDelegate(Delegate.GPU)
      .build()
    val opts = HolisticLandmarker.HolisticLandmarkerOptions.builder()
      .setBaseOptions(base)
      .setRunningMode(RunningMode.VIDEO)
      // Not part of this pipeline, and both cost time on every frame.
      .setOutputFaceBlendshapes(false)
      .setOutputPoseSegmentationMasks(false)
      .build()
    HolisticLandmarker.createFromOptions(proxy.context, opts)
  } catch (e: Throwable) {
    Log.e(TAG, "failed to create HolisticLandmarker", e)
    null
  }

  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
    val detector = landmarker ?: return null
    val timestampMs = (arguments?.get("timestampMs") as? Number)?.toLong() ?: 0L

    val result: HolisticLandmarkerResult = try {
      // frame.imageProxy is a CameraX type this module does not have on its
      // compile classpath, and ImageProxy.toBitmap() is not available in every
      // CameraX version. frame.image is android.media.Image - a framework class
      // - so it needs no extra dependency. The camera is configured for RGBA
      // output (pixelFormat="rgb" in LandmarkCamera), so plane 0 can be copied
      // straight into a bitmap without a YUV conversion.
      val bitmap = frame.image.toArgbBitmap() ?: return null
      detector.detectForVideo(BitmapImageBuilder(bitmap).build(), timestampMs)
    } catch (e: Throwable) {
      Log.w(TAG, "holistic detection failed for one frame", e)
      return null
    }

    val streams = listOf(
      result.faceLandmarks(),
      result.poseLandmarks(),
      result.leftHandLandmarks(),
      result.rightHandLandmarks(),
    )

    val totalFloats = HEADER_FLOATS + streams.sumOf { it.size } * FLOATS_PER_LANDMARK
    val byteCount = totalFloats * 4

    val shared = SharedArray(proxy, byteCount)
    val buffer: ByteBuffer = shared.byteBuffer.order(ByteOrder.nativeOrder())
    val floats = buffer.asFloatBuffer()

    floats.put(0, SCHEMA_VERSION)
    floats.put(1, timestampMs.toFloat())
    streams.forEachIndexed { i, stream -> floats.put(2 + i, stream.size.toFloat()) }

    var offset = HEADER_FLOATS
    for (stream in streams) {
      for (point in stream) {
        floats.put(offset, point.x())
        floats.put(offset + 1, point.y())
        floats.put(offset + 2, point.z())
        floats.put(offset + 3, point.visibility().orElse(1f))
        offset += FLOATS_PER_LANDMARK
      }
    }

    return shared
  }

  /**
   * Copies an RGBA frame into an ARGB_8888 bitmap.
   *
   * Rows can be padded, so a plane whose stride exceeds width * 4 is copied row
   * by row rather than in one block - copying the padding straight through
   * would shear the image and produce landmarks that are subtly wrong rather
   * than obviously broken.
   */
  private fun Image.toArgbBitmap(): Bitmap? {
    val plane = planes.firstOrNull() ?: return null
    val buffer = plane.buffer.also { it.rewind() }
    val rowStride = plane.rowStride
    val tightStride = width * BYTES_PER_PIXEL

    val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
    if (rowStride == tightStride) {
      bitmap.copyPixelsFromBuffer(buffer)
      return bitmap
    }

    val packed = ByteBuffer.allocateDirect(tightStride * height).order(ByteOrder.nativeOrder())
    val row = ByteArray(tightStride)
    for (y in 0 until height) {
      buffer.position(y * rowStride)
      buffer.get(row, 0, tightStride)
      packed.put(row)
    }
    packed.rewind()
    bitmap.copyPixelsFromBuffer(packed)
    return bitmap
  }
}

File: native/holistic/ios/HolisticFrameProcessorPlugin.swift
import Foundation
import MediaPipeTasksVision
import VisionCamera

/**
 * Runs MediaPipe HolisticLandmarker over each camera frame and returns one
 * packed Float32 buffer, base64-encoded.
 *
 * The layout is specified and unit-tested in `lib/extractors/holistic-buffer.ts`,
 * which is the authority for this encoding:
 *   [0]     schema version (1)
 *   [1]     timestamp ms since capture start
 *   [2..5]  face / pose / leftHand / rightHand counts, 0 when undetected
 *   then    each stream in that order, four floats per landmark: x, y, z, visibility
 *
 * A packed buffer is used rather than a dictionary because a holistic result is
 * ~543 landmarks; bridging that as nested objects every frame would dominate the
 * frame budget.
 *
 * It is returned as base64 rather than a SharedArray because the result is
 * handed to JS through Worklets.createRunOnJS, whose shared-value converter
 * rejects ArrayBuffers and wraps arrays element by element. A string crosses
 * as one copy. Bytes are little-endian, matching the reader; every Apple
 * platform this ships to is little-endian, and the conversion below is
 * explicit rather than relying on that.
 *
 * Nothing here retains, encodes, or writes a frame. The sample buffer is read
 * within the call and released by VisionCamera.
 */
@objc(HolisticFrameProcessorPlugin)
public class HolisticFrameProcessorPlugin: FrameProcessorPlugin {
  private static let schemaVersion: Float = 1
  private static let headerFloats = 6
  private static let floatsPerLandmark = 4

  private let landmarker: HolisticLandmarker?

  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = nil) {
    self.landmarker = Self.makeLandmarker()
    super.init(proxy: proxy, options: options)
  }

  private static func makeLandmarker() -> HolisticLandmarker? {
    guard let modelPath = Bundle.main.path(
      forResource: "holistic_landmarker", ofType: "task"
    ) else {
      NSLog("[holistic] holistic_landmarker.task is missing from the app bundle")
      return nil
    }

    let opts = HolisticLandmarkerOptions()
    opts.baseOptions.modelAssetPath = modelPath
    opts.runningMode = .video
    // Blendshapes and segmentation masks are not part of this pipeline and cost
    // time on every frame, so they stay off.
    opts.outputFaceBlendshapes = false
    opts.outputPoseSegmentationMasks = false

    do {
      return try HolisticLandmarker(options: opts)
    } catch {
      NSLog("[holistic] failed to create HolisticLandmarker: \(error)")
      return nil
    }
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    guard let landmarker = self.landmarker else { return nil }

    let timestampMs = (arguments?["timestampMs"] as? NSNumber)?.doubleValue ?? 0

    guard let image = try? MPImage(sampleBuffer: frame.buffer),
          let result = try? landmarker.detect(
            videoFrame: image,
            timestampInMilliseconds: Int(timestampMs)
          )
    else {
      return nil
    }

    let streams = [
      result.faceLandmarks,
      result.poseLandmarks,
      result.leftHandLandmarks,
      result.rightHandLandmarks,
    ]

    let totalFloats = Self.headerFloats
      + streams.reduce(0) { $0 + $1.count } * Self.floatsPerLandmark

    var packed = [Float](repeating: 0, count: totalFloats)
    packed[0] = Self.schemaVersion
    packed[1] = Float(timestampMs)
    for (i, stream) in streams.enumerated() {
      packed[2 + i] = Float(stream.count)
    }

    var offset = Self.headerFloats
    for stream in streams {
      for point in stream {
        packed[offset] = point.x
        packed[offset + 1] = point.y
        packed[offset + 2] = point.z
        packed[offset + 3] = point.visibility?.floatValue ?? 1
        offset += Self.floatsPerLandmark
      }
    }

    var bytes = Data(capacity: totalFloats * MemoryLayout<UInt32>.size)
    for value in packed {
      // littleEndian is a no-op on Apple silicon and Intel, but stating it
      // keeps the byte order a property of the format rather than the host.
      withUnsafeBytes(of: value.bitPattern.littleEndian) { bytes.append(contentsOf: $0) }
    }
    return bytes.base64EncodedString()
  }
}
